pkg/didcomm/crypto/authcrypt/encrypt.go

```diff
@@ -0,0 +1,288 @@
+	"golang.org/x/crypto/chacha20poly1305"
+	"golang.org/x/crypto/poly1305"
+)
+
+// This package deals with Authcrypt encryption for Packing/Unpacking DID Comm exchange
+// Chacha20Poly35 encryption/authentication
+
+// RandReader is a cryptographically secure random number generator (stubbed out in tests).
+var RandReader = rand.Reader
+
+// Crypter represents an Authcrypt Encrypter (Decrypter) that outputs/reads JWE envelopes
+type Crypter struct {
+	Sender     interface{}
+	Recipients []interface{}
+	Alg        string
+	NonceSize  int
+}
+
+// JweEnvelope represents a JWE envelope as per the Aries Encryption envelop specs
+type JweEnvelope struct {
+	Protected  jweHeaders  `json:"Protected,omitempty"`
+	Recipients []Recipient `json:"recipients,omitempty"`
+	Aad        string      `json:"aad,omitempty"`
+	Iv         string      `json:"iv,omitempty"`
+	Tag        string      `json:"tag,omitempty"`
+	CipherText string      `json:"ciphertext,omitempty"`
+}
+
+// jweHeaders are the Protected JWE headers in a map format
+type jweHeaders map[string]string
+
+// Recipient is a recipient of an envelop including the shared encryption key
+type Recipient struct {
+	EncryptedKey string           `json:"encrypted_key,omitempty"`
+	Header       RecipientHeaders `json:"header,omitempty"`
+}
+
+// RecipientHeaders are the recipient headers
+type RecipientHeaders struct {
+	Apu string `json:"apu,omitempty"`
+	Iv  string `json:"iv,omitempty"`
+	Tag string `json:"tag,omitempty"`
+	Kid string `json:"kid,omitempty"`
+	Oid string `json:"oid,omitempty"`
+}
+
+// NewAuthCrypter will create an encrypter instance to 'AuthCrypt' payloads for the given sender and recipients arguments
+// and the encryption alg argument. Possible algorithms supported are:
+// C20P (chacha20-poly1035 ietf)
+// XC20P (xchacha20-poly1035 ietf)
+func NewAuthCrypter(sender interface{}, recipients []interface{}, alg string) (*Crypter, error) {
+	var nonceSize int
+	switch alg {
+	case "C20P":
+		nonceSize = chacha20poly1305.NonceSize
+	case "XC20P":
+		nonceSize = chacha20poly1305.NonceSizeX
+	default:
+		return nil, errors.New(fmt.Sprintf("Alg %s not supported", alg))
+	}
+
+	c := &Crypter{
+		sender,
+		recipients,
+		alg,
+		nonceSize,
+	}
+
+	typeofS := reflect.TypeOf(sender)
+	if typeofS != reflect.TypeOf(&ecdsa.PrivateKey{}) {
+		return nil, errors.New("sender key not supported, it is not an EC private key")
+	}
+
+	for i, recipient := range recipients {
+		typeofR := reflect.TypeOf(recipient)
+		if typeofR != reflect.TypeOf(&ecdsa.PublicKey{}) {
+			return nil, errors.New(fmt.Sprintf("recipient %d key not supported, it is not an EC public key", i+1))
+		}
+	}
+
+	return c, nil
+}
+
+// Encrypt will JWE encode the payload argument for the sender and recipients
+// Using (X)Chacha20Poly1035 encryption & authentication (Authcrypt)
+func (c *Crypter) Encrypt(payload string) (string, error) {
+	headers := jweHeaders{
+		"typ": "prs.hyperledger.aries-auth-message",
+		"alg": "ECDH-SS+XC20PKW",
```

Comment:
This isn't the `alg` I see defined [in the RFC](https://tools.ietf.org/html/draft-amringer-jose-chacha-01#section-3)

Reply:
according to @gamringer, this encryption is static-static

Reply:
> Test coverage is a bit low.

the remaining uncovered lines are related to JSON marshal errors, or derive keys.. I will try to increase the coverage a bit.

Reply:
alg as per the wip draft discussed with @gamringer:
```
Baha-sk
12:23 PM
another question.. is this right ECDH-SS+XC20PKW or you meant ECDH-ES+XC20PKW in the headers?
gamringer
1:26 PM
I did mean ECDH-SS+XC20PKW, because with ECDH-ES*, you have to put the whole ephemeral public key in the header, whereas the requirement for authcrypt is to have only a key identifier
```
so the draft for this JWE will use SS in the algorithm and should be updated.